device/cuda/src/utils/sync.cuh

```diff
@@ -0,0 +1,40 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+namespace traccc::cuda {
+/**
+ * @brief Calculate the index of the current thread in the set of threads in
+ * the warp for which a predicate evaluates to true.
+ *
+ * @param predicate The predicate to evaluate.
+ *
+ * @returns The tuple (T, I), where T is the total number of threads in the
+ * warp for which the provided predicate was true, and where I is a unique,
+ * consecutive index in the set of threads for which this was true.
+ *
+ * @example If a predicate P evaluates as `[T, F, F, T]` in a 4-lane warp, the
+ * call to `warp_indexed_ballot_sync(P)` will evaluate to the values
+ * `[(2, 0), (2, ?), (2, ?), (2, 1)]` where `?` indicates an undefined value.
+ *
+ * @warning The value of T is always well-defined, the value if I is only
+ * well-defined if the given predicate was true for the given thread.
+ *
+ * @note This function forces thread synchronization.
+ */
+__device__ __forceinline__ std::pair<uint32_t, uint32_t>
+warp_indexed_ballot_sync(bool predicate, uint32_t active = 0xFFFFFFFFu) {
+    uint32_t mask = __ballot_sync(active, predicate);
+
+    uint32_t tot = __popc(mask);
+    uint32_t idx =
+        __popc(mask & ~(0xFFFFFFFEu << (threadIdx.x % warpSize))) - 1u;
```

Comment:
For sure, I can add a comment. If you are interested, the line starts out with the mask returned by the ballot synchronization. This means that it is a 32-bit number where bit _n_ is true if and only if thread _n_ voted yes.

So, for example, if we were to vote on whether the thread index was a power of 2, or 3 times a power of 2 ([A029744](https://oeis.org/A029744)), it might look something like this, where the top rows indicate the bit number (counting right to left):

```
 3         2         1         0
10987654321098765432109876543210
00000001000000010001000101011110
```

Now, the list of threads which voted true, counting upwards, is [1, 2, 3, 4, 6, 8, 12, 16, 24]. For each of these threads, we want the index to be their index in that list. So thread 1 would get index 0, thread 2 would get index 1, thread 8 would get index 5, and so forth.

To do this, we can count the number of true bits set to one in the voted mask for which the index is equal to or smaller than the thread index. For example, thread 8 would look at the following bits (underlined):

```
 3         2         1         0
10987654321098765432109876543210
00000001000000010001000101011110
                        ________
```

There are five ones in the underlined region, and so the index is 5 (as we discussed above)!

The expression `~(0xFFFFFFFEu << (threadIdx.x % warpSize))` generates, in effect, the underlined region. We start out with a binary number that is all ones:

```
 3         2         1         0
10987654321098765432109876543210
11111111111111111111111111111111
```

Then we shift it left equal to the thread index, throwing away bits that are on the left, and padding the right with zeroes:

```
 3         2         1         0
10987654321098765432109876543210
11111111111111111111111100000000
```

We then take the inverse of this flag:

```
 3         2         1         0
10987654321098765432109876543210
00000000000000000000000011111111
```

And finally, we take the bitwise and of this and the original mask:

```
 3         2         1         0
10987654321098765432109876543210
00000001000000010001000101011110
00000000000000000000000011111111 &
--------------------------------
00000000000000000000000001011110 =
```

Finally, it runs a _population count_ operation, which simply counts the number of ones in the binary expansion of a number. Hopefully this helps clear up why this code works. :smile:

Reply:
Thanks by the way, writing it down like this made me realize that I made a very small inefficiency mistake, which I have now fixed.

Reply:
Thanks for the thorough explanation, it was very clear!